src/lib/edf_schedule.c

```diff
@@ -540,6 +540,31 @@ static void schedule_edf_task_free(struct task *task)
 	edf_sch_set_pdata(task, NULL);
 }
 
+/* scheduler of idle tasks (non prioritized) */
+void schedule_idle(void)
+{
+	struct edf_schedule_data *sch =
+		(*arch_schedule_get_data())->edf_sch_data;
+	struct list_item *tlist;
+	struct task *task;
+	struct list_item *clist;
+
+	/* first check if any higher priority task is waiting */
+	schedule();
+
```

Comment:
Maybe even better idea would be to not add any additional `schedule_idle()` function, but rather execute below code in normal `schedule()` and check for `interrupt_level == 0`.

Reply:
I think we can not run only idle tasks here but both, idle and prioritized. My concern is long last IRQ handlers and/or missed tasks deadline. I mean, if we decide to invoke queued task only during scheduling then some tasks may miss their deadline because we check queued list only in two places - during scheduling of new tasks and here in idle loop.

Reply:
I was thinking about something like this:

`static void schedule_edf(void)
{
	struct edf_schedule_data *sch =
		(*arch_schedule_get_data())->edf_sch_data;
	struct list_item *tlist;
	struct list_item *clist;
	struct task *edf_task;

	uint32_t flags;

	tracev_edf_sch("schedule_edf()");

	spin_lock_irq(&sch->lock, flags);

	/* make sure we have a queued task in the list first before we
	   start scheduling as contexts switches are not free. */
	list_for_item(tlist, &sch->list) {
		edf_task = container_of(tlist, struct task, list);

		/* schedule if we find any queued tasks */
		if (edf_task->state == SOF_TASK_STATE_QUEUED) {
			spin_unlock_irq(&sch->lock, flags);
			/* the scheduler is run in IRQ context */
			interrupt_set(PLATFORM_SCHEDULE_IRQ);
			goto idle;
		}
	}

	/* no task to schedule */
	spin_unlock_irq(&sch->lock, flags);

idle:
	if (arch_interrupt_get_level() == 0) {
		list_for_item_safe(clist, tlist, &sch->idle_list) {
			edf_task = container_of(clist, struct task, list);

			/* run task if we find any queued */
			if (edf_task->state == SOF_TASK_STATE_QUEUED) {
				edf_task->func(edf_task->data);
				edf_task->state = SOF_TASK_STATE_COMPLETED;

				/* task done, remove it from the list */
				list_item_del(clist);
			}
		}
	}
}`

Reply:
Yes, this is what I was thinking about, please check updated version.